src/imageops/mod.rs

```diff
@@ -190,6 +192,86 @@ where
     }
 }
 
+/// Fill the image with a linear vertical gradient
+/// 
+/// # Examples
+/// ```no_run
+/// use image::{Rgba, RgbaImage, Pixel};
+/// 
+/// fn main() {
+///     let mut img = RgbaImage::new(100, 100);
+///     let start = Rgba::from_slice(&[0, 128, 0, 0]);
+///     let end = Rgba::from_slice(&[255, 255, 255, 255]);
+/// 
+///     image::imageops::vertical_gradient(&mut img, start, end);
+///     img.save("vertical_gradient.png").unwrap();
+/// }
+pub fn vertical_gradient<S, P, I>(img: &mut I, start: &P, stop: &P)
+where
+    I: GenericImage<Pixel = P>,
+    P: Pixel<Subpixel = S> + 'static,
+    S: Primitive + Enlargeable + 'static,
+{
+    for y in 0..img.height() {
+        let pixel = start.map2(stop, |a, b| {
+            let a = <S::Larger as NumCast>::from(a).unwrap();
+            let b = <S::Larger as NumCast>::from(b).unwrap();
+            let y = <S::Larger as NumCast>::from(y).unwrap();
+            let height = <S::Larger as NumCast>::from(img.height()).unwrap();
+
+            let min = cmp::min(a, b);
+            let max = cmp::max(a, b);
+
+            let res = min + (y * (max - min) / height);
```

Comment:
This was completely wrong indeed. Tried to reverse start and stop color in the example and got the same output. I am however struggling to use negative values in the computation as the `Larger` type is unsigned and cannot go under zero:

```
thread 'main' panicked at 'attempt to subtract with overflow', /home/microjoe/.rustup/toolchains/nightly-x86_64-unknown-linux-gnu/lib/rustlib/src/rust/src/libcore/ops/arith.rs:194:45
note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace
```

I introduced a `SignedLarger` type for this kind of needs, but I am not sure this will introduce breaking changes or not. With this change I can reverse the gradient colors just fine.

Reply:
This was completely wrong indeed. Reversing start and stop color did not produce expected changes. I have introduced a new `SignedLarger` type in order to help with negative values handling. Not sure if this is a breaking change or if you have other implementation proposal.

I was looking at something like `<S::Larger as SignedNumCast>` to find the signed type of "u32" to be "i32" etc. but did not found what I was looking for, hence this proposal.

Reply:
I see. That means that `Enlargable` isn't the right trait afterall. At the very minimum you need types with an equivalent of a [`lerp`](https://en.cppreference.com/w/cpp/numeric/lerp) method. What about introducing a new trait for this?

Reply:
Done. I am afraid that while the "integer" approach works, maybe using floating points should be used. There is a risk of panic if the lerp is used on a huge "i" (and "n") values. Is introducing floating points worth the trouble?

Note that the name is probably wrong -- as "official" `lerp` method seems to be planned: https://github.com/rust-lang/rust/pull/71016